File: app/models/task.py
"""
Task model — collaborative task management with assignments.
"""
from sqlalchemy import Column, Integer, BigInteger, String, Text, DateTime, Enum, ForeignKey
from sqlalchemy.sql import func
from app.core.database import Base
import enum


class TaskStatus(str, enum.Enum):
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    DONE = "done"


class TaskPriority(str, enum.Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class Task(Base):
    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(500), nullable=False)
    description = Column(Text, nullable=True)
    status = Column(Enum(TaskStatus), default=TaskStatus.TODO, nullable=False)
    priority = Column(Enum(TaskPriority), default=TaskPriority.MEDIUM, nullable=False)
    label = Column(String(100), nullable=True)
    category = Column(String(100), nullable=True)      # e.g., "GDT", "Kado24", "Personal"
    subcategory = Column(String(100), nullable=True)    # e.g., "ETL", "Frontend", "Learning"

    # Task Group hierarchy
    group_id = Column(Integer, ForeignKey("task_groups.id", ondelete="SET NULL"), nullable=True)
    subgroup_id = Column(Integer, ForeignKey("task_subgroups.id", ondelete="SET NULL"), nullable=True)

    # Creator and assignee (Telegram IDs)
    creator_id = Column(BigInteger, nullable=False)
    creator_name = Column(String(255), nullable=True)
    assignee_id = Column(BigInteger, nullable=True)
    assignee_name = Column(String(255), nullable=True)

    # Source context
    source_chat_id = Column(BigInteger, nullable=True)
    source_message_id = Column(BigInteger, nullable=True)

    due_date = Column(DateTime(timezone=True), nullable=True)
    estimated_hours = Column(Float, nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    version = Column(Integer, default=1)
    last_modified_by = Column(String(255), nullable=True)

    def __repr__(self):
        return f"<Task(id={self.id}, title={self.title[:30]}, status={self.status})>"

File: mobile/lib/models/task.dart
class Task {
  final int id;
  final String title;
  final String? description;
  final String status;
  final String priority;
  final String? assignee;
  final String? category;
  final String? subcategory;
  final String? dueDate;
  final String? createdAt;
  final int? groupId;
  final String? groupName;

  Task({
    required this.id, required this.title, this.description,
    required this.status, required this.priority, this.assignee,
    this.category, this.subcategory, this.dueDate, this.createdAt,
    this.groupId, this.groupName,
  });

  factory Task.fromJson(Map<String, dynamic> json) => Task(
    id: json['id'] ?? 0,
    title: json['title'] ?? '',
    description: json['description'],
    status: json['status'] ?? 'todo',
    priority: json['priority'] ?? 'medium',
    assignee: json['assignee'] ?? json['assignee_name'],
    category: json['category'],
    subcategory: json['subcategory'],
    dueDate: json['due_date'],
    createdAt: json['created_at'],
    groupId: json['group_id'],
    groupName: json['group_name'],
  );

  Map<String, dynamic> toJson() => {
    'title': title, 'description': description, 'status': status,
    'priority': priority, 'assignee_name': assignee, 'category': category,
    'subcategory': subcategory, 'due_date': dueDate,
    if (groupId != null) 'group_id': groupId,
  };
}
